### PR TITLE
feat: simulate swap as part of quotes

### PR DESCRIPTION
**Pull request by p0mvn**
https://osmosis-network.slack.com/archives/C079LNKRUCB/p1730651766930679

Implementing `MsgSwapExactAmountIn` quote estimation logic directly into SQS.

This will allow to wire SQS integration into Polaris. Ref: https://github.com/polaris-portal/polaris/pull/1011/files

Currently, in Polaris, as we get a quote, we also want to know the pricing for it.

To avoid round trips, we can have an optional query parameter to request SQS to run simulation logic immediately after estimating a quote.

## Support

- "out given in", direct (non-split) quotes
- controlled by the optiona `simulationAddress` and `simulationSlippageTolerance` query parameters
- backwards compatible

## Testing

- Unit tests with mock
- Integration test

## Release Notes

- **New Features**
  - Introduced a `PriceMsgs` method for improved gas estimation and fee calculation in message simulations.
  - Added a `QuoteSimulator` interface and a `SimulateQuote` method for simulating quotes in token swaps.
  - Enhanced `GetQuoteRequest` to handle simulation parameters, including `SimulatorAddress` and `SlippageToleranceMultiplier`.
  - Added a mock implementation of the `QuoteSimulator` for testing purposes.

- **Improvements**
  - Updated `RouterHandler` to utilize the new quote simulation functionality.
  - Added a `SetQuotePriceInfo` method to manage quote price data effectively.
  - Enhanced server initialization to integrate the new quote simulator.

- **Bug Fixes**
  - Improved error handling in the quote simulation process to ensure proper validation and responses.

## Release Notes

- **New Features**
  - Introduced `PriceMsgs` method for improved gas estimation and fee calculation in message simulation.
  - Added `QuoteSimulator` interface and `SimulateQuote` method for simulating quotes with slippage tolerance.
  - Enhanced `GetOptimalQuote` method to support quote simulation based on a specified address and slippage tolerance.
  - New fields in `GetQuoteRequest` to manage simulation parameters.
  - Added optional `price_info` parameter to `QuoteExactAmountInResponse` for additional pricing details.

- **Bug Fixes**
  - Adjusted error handling for quote simulation to ensure proper responses.

- **Documentation**
  - Updated Swagger documentation to include new parameters for quote simulation.

- **Tests**
  - Added tests for simulation slippage tolerance and updated existing tests to accommodate new parameters.

### CODE REVIEW

File: domain/quote_simulator.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/osmomath"
+)
+
+// QuoteSimulator simulates a quote and returns the gas adjusted amount and the fee coin.
+type QuoteSimulator interface {
+	// SimulateQuote simulates a quote and returns the gas adjusted amount and the fee coin.
+	// CONTRACT:
+	// - Only direct (non-split) quotes are supported.
```
Comment (deividaspetraitis) [inline domain/quote_simulator.go:14]: Why only non-split routes are supported? Nit: typo in `Retursn`

File: quotesimulator/quote_simulator.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+			TokenOutDenom: r.GetTokenOutDenom(),
+		}
+	}
+
+	// Slippage bound from the token in and provided slippage tolerance multiplier
+	tokenOutAmt := quote.GetAmountOut()
+	slippageBound := tokenOutAmt.ToLegacyDec().Mul(slippageToleranceMultiplier).TruncateInt()
+
+	// Create the swap message
+	swapMsg := &poolmanagertypes.MsgSwapExactAmountIn{
```
Comment (deividaspetraitis) [inline quotesimulator/quote_simulator.go:60]: Can it be that result can vary depending of swap method ( `poolmanagertypes.MsgSwapExactAmountIn` vs  `poolmanagertypes.MsgSwapExactAmountOut` ) for some cases such as imbalanced pools, etc?

**Review by deividaspetraitis — APPROVED** [review]

File: domain/quote_simulator.go (hunk 1 above)
Comment (p0mvn) [reply domain/quote_simulator.go:14]: Splits require a separate message and increase the complexity drastically.

My ultimate goal is to integrate SQS into Polaris at the lowest cost, and then we can iterate on improvements, including adding splits

File: quotesimulator/quote_simulator.go (hunk 2 above)
Comment (p0mvn) [reply quotesimulator/quote_simulator.go:60]: "in given out" is not supported in simulation for the same reason I mentioned here:
https://github.com/osmosis-labs/sqs/pull/547#discussion_r1828122218

**Pull request merged**